source/monitoring/cluster-monitoring.txt

```diff
@@ -0,0 +1,136 @@
+.. _php-cluster-monitoring:
+
+==================
+Cluster Monitoring
+==================
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: code example, server, topology
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecols
+
+Overview
+--------
+
+This guide shows you how to use the {+php-library+} to monitor topology
+events in a MongoDB instance, replica set, or sharded cluster. The
+driver creates topology events, also known as Server Discovery and
+Monitoring (SDAM) events, when there are any changes in the state of the
+MongoDB instance or cluster that you are connected to.
+
+You might use information about topology events in your
+application to understand cluster changes, assess cluster health, or
+perform capacity planning.
+
+Subscribe to Events
+-------------------
+
+You can access details about SDAM events by subscribing to them
+in your application. To subscribe to an event, create a class that
+implements the ``MongoDB\Driver\Monitoring\SDAMSubscriber`` interface,
+then use the ``MongoDB\Client::addSubscriber()`` method to register the
+event subscriber with your ``MongoDB\Client`` instance.
+
+The following code creates the ``MySubscriber`` class, which implements
+the ``SDAMSubscriber`` interface. The class is defined with a method to
+output a message when a ``ServerOpeningEvent`` is generated by the
+server:
+
+.. literalinclude:: /includes/monitoring/sdam.php
+   :start-after: start-mysubscriber
+   :end-before: end-mysubscriber
+   :language: php
+   :copyable:
+   :dedent:
+
+.. note::
+
+   As shown in the preceding code, you must implement all of the methods
+   of the ``SDAMSubscriber`` interface, even for events you are not subscribing to.
+   You can implement these methods with empty bodies so that the library
+   does not generate any messages for these events.
+
+Then, use the ``addSubscriber()`` method to register ``MySubscriber``
+with the client, as shown in the following code:
+
+.. literalinclude:: /includes/monitoring/sdam.php
+   :start-after: start-add-sub
+   :end-before: end-add-sub
+   :language: php
+   :copyable:
+   :dedent:
+
+When you run the application, your subscriber records the SDAM event and
+outputs messages such as the following:
+
+.. code-block:: none
+   :copyable: false
+
+   Server opening on ac-rmuag0v-shard-00-00.gh0qg50.mongodb.net:27017
+   Server opening on ac-rmuag0v-shard-00-01.gh0qg50.mongodb.net:27017
+   Server opening on ac-rmuag0v-shard-00-02.gh0qg50.mongodb.net:27017
+
+Event Descriptions
+------------------
+
+You can subscribe to the following SDAM events by implementing the
+corresponding method from the ``SDAMSubscriber`` interface:
```

Comment:
This talks about the "corresponding method" from [SDAMSubscriber](https://www.php.net/manual/en/class.mongodb-driver-monitoring-sdamsubscriber.php), but those names aren't presented in the table below. It would be a bit redundant to dedicate a column to the method name, but I think some links to the PHP.net docs could make this a more helpful resource.

Consider working in links to the following:

 * The [SDAMSubscriber](https://www.php.net/manual/en/class.mongodb-driver-monitoring-sdamsubscriber.php) interface.
 * Individual class pages for each event under https://www.php.net/manual/en/mongodb.monitoring.php. For example, ServerOpeningEvent can link to https://www.php.net/manual/en/class.mongodb-driver-monitoring-serveropeningevent.php.

Reply:
The API doc for SDAMSubscriber is linked at the end of the page, but I will link each event in the table.